app/src/main/java/ciserver/ContinuousIntegrationServer.java

```diff
@@ -51,6 +44,33 @@ private void handlePushEvent(PushEvent event) {
         } catch (Exception e) {
             System.err.println(e);
         }
+    }
+
+    private PipelineStatus executePipeline(PushEvent event, Pipeline pipeline, PipelineObserver observer) {
+        pipeline.addObserver(observer);
+        return pipeline.start(TargetStage.ALL);
+
+    }
+
+    private void handlePushEvent(PushEvent event) {
+        System.err.printf("%s, %s", event.ref, event.headCommit.url);
+
+        var pipeline = new Pipeline(event, "../pipeline");
+        var status = executePipeline(event, pipeline, new PipelineObserver() {
+            @Override
+            public void update(TargetStage stage, PipelineStatus status) {
+                System.err.printf("\tRunning stage: %s: %s\n", stage, status);
+
+                if (status == PipelineStatus.Fail) {
+                    sendUpdateRequest(event, CommitStatus.FAILURE,
+                            String.format("Failed during stage %s with status %s.", stage, status), stage);
+                }
+            }
+        });
```

Comment:
After having thought about it a bit I have come to the conclusion that `executePipeline` is unnecessary. See suggestion without bellow.

```suggestion
 Pipeline pipeline = new Pipeline(event, "../pipeline");
        pipeline.addObserver(new PipelineObserver() {
            @Override
            public void update(TargetStage stage, PipelineStatus status) {
                System.err.printf("\tRunning stage: %s: %s\n", stage, status);

                if (status == PipelineStatus.Fail) {
                    sendUpdateRequest(
                            event,
                            CommitStatus.FAILURE,
                            String.format("Failed during stage %s with status %s.", stage, status),
                            stage
                    );
                }
            }
        });

        var status = pipeline.start(TargetStage.ALL);
```

Reply:
Agreed; if I'm not mistaken it just adds the observer and calls `start()` - nothing I see needs to be in a separate function, especially as it's only called once (for the one pipeline we have). If we would want to `start()` multiple instances at once in the future this could be reworked, but for now, I think it's fine doing it as in this suggestion.

Reply:
The purpose of the detachment is to effectively test the function.

Reply:
It's called from the test.

Reply:
But does the `executePipeline()` function really need testing? Testing it only tests that `pipeline.addObserver()` and `pipeline.start()` work correctly - something better done in separate tests anyway imo. And any test that runs `executePipeline()` as part of it may as well just run the two lines of code, right?

Reply:
I still think this is the only test we can do of this, in the time frame we have.